repose-aggregator/services/atom-feed/impl/src/main/scala/org/openrepose/nodeservice/atomfeed/impl/actors/FeedReader.scala

```diff
@@ -0,0 +1,149 @@
+import org.openrepose.nodeservice.atomfeed.impl.AtomEntryStreamBuilder.AuthenticationException
+import org.openrepose.nodeservice.atomfeed.impl.actors.Notifier.{FeedReaderActivated, FeedReaderCreated, FeedReaderDeactivated, FeedReaderDestroyed}
+import org.openrepose.nodeservice.atomfeed.impl.actors.NotifierManager._
+import org.openrepose.nodeservice.atomfeed.impl.auth.AuthenticationRequestContextImpl
+import org.slf4j.MDC
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+object FeedReader {
+
+  def props(feedUri: String,
+            authenticatedRequestFactory: Option[AuthenticatedRequestFactory],
+            notifierManager: ActorRef,
+            pollingFrequency: Int,
+            order: EntryOrderType,
+            reposeVersion: String): Props =
+    Props(new FeedReader(feedUri, authenticatedRequestFactory, notifierManager, pollingFrequency, order, reposeVersion))
+
+  object ReadFeed
+
+  object ScheduleReading
+
+  object CancelScheduledReading
+
+}
+
+class FeedReader(feedUri: String,
+                 authenticatedRequestFactory: Option[AuthenticatedRequestFactory],
+                 notifierManager: ActorRef,
+                 pollingFrequency: Int,
+                 order: EntryOrderType,
+                 reposeVersion: String)
+  extends Actor with LazyLogging {
+
+  import FeedReader._
+
+  val parser = Abdera.getInstance().getParser
+  val feedUrl = new URL(feedUri)
+
+  var firstReadDone = false
+  var highWaterMark: Option[IRI] = None
+  var schedule: Option[Cancellable] = None
+
+  override def preStart(): Unit = {
+    // bind this feed reader to its corresponding notifier manager
+    notifierManager ! BindFeedReader(self)
+    notifierManager ! FeedReaderCreated
+  }
+
+  override def postStop(): Unit = {
+    notifierManager ! FeedReaderDestroyed
+  }
+
+  override def receive: Receive = {
+    case ReadFeed =>
+      val requestId = java.util.UUID.randomUUID().toString
+      MDC.put(TracingKey.TRACING_KEY, requestId)
+
+      try {
+        val entryStream = AtomEntryStreamBuilder.build(feedUrl, AuthenticationRequestContextImpl(reposeVersion, requestId), authenticatedRequestFactory)
+
+        if (firstReadDone) {
+          val newEntryStream = highWaterMark match {
+            case Some(mark) => entryStream.takeWhile(entry => !entry.getId.equals(mark))
+            case None => entryStream
+          }
+
+          newEntryStream foreach { entry =>
+            val entryString = new StringWriter()
+            entry.writeTo(entryString)
+            // todo: wrap atom entry in type-safe object
+            notifierManager ! Notify(entryString.toString)
+          }
+        }
+
+        firstReadDone = true
+        highWaterMark = entryStream.headOption.map(_.getId)
+      } catch {
+        // fixme: Is there any case where this feed reader actor should be killed? Or restarted after some time?
```

Comment:
This is the fixme I referred to in a previous comment. Currently the feed reader actor will just keep on chugging if an exception occurs. One example of where we might not want this is when user credentials are bad, and the authentication component keeps failing (throwing an exception). Every x seconds, this actor will log the error, but it won't turn itself off or anything.